cudax/include/cuda/experimental/__memory_resource/async_memory_resource.cuh

```diff
@@ -0,0 +1,312 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDA Experimental in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDAX__MEMORY_RESOURCE_CUDA_ASYNC_MEMORY_RESOURCE_H
+#define _CUDAX__MEMORY_RESOURCE_CUDA_ASYNC_MEMORY_RESOURCE_H
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+// cudaMallocAsync was introduced in CTK 11.2
+#if !defined(_CCCL_COMPILER_MSVC_2017) && !defined(_CCCL_CUDACC_BELOW_11_2)
+
+#  if !defined(_CCCL_CUDA_COMPILER_NVCC) && !defined(_CCCL_CUDA_COMPILER_NVHPC)
+#    include <cuda_runtime.h>
+#    include <cuda_runtime_api.h>
+#  endif // !_CCCL_CUDA_COMPILER_NVCC && !_CCCL_CUDA_COMPILER_NVHPC
+
+#  include <cuda/__memory_resource/get_property.h>
+#  include <cuda/__memory_resource/properties.h>
+#  include <cuda/__memory_resource/resource_ref.h>
+#  include <cuda/std/__concepts/__concept_macros.h>
+#  include <cuda/std/__cuda/api_wrapper.h>
+#  include <cuda/std/__new_>
+#  include <cuda/std/cstddef>
+#  include <cuda/stream_ref>
+
+#  include <cuda/experimental/__device/device_ref.cuh>
+#  include <cuda/experimental/__memory_resource/async_memory_pool.cuh>
+#  include <cuda/experimental/__stream/stream.cuh>
+
+#  if _CCCL_STD_VER >= 2014
+
+namespace cuda::experimental::mr
+{
+
+//! @brief global stream to synchronize in the synchronous interface of \c async_memory_resource
+inline ::cuda::stream_ref __async_memory_resource_sync_stream()
+{
+  static ::cuda::experimental::stream __stream{};
+  return __stream;
+}
+
+//! @rst
+//! .. _cudax-memory-resource-async:
+//!
+//! Stream ordered memory resource
+//! ------------------------------
+//!
+//! ``async_memory_resource`` uses `cudaMallocFromPoolAsync / cudaFreeAsync
+//! <https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY__POOLS.html>`__ for allocation/deallocation. A
+//! ``async_memory_resource`` is a thin wrapper around a \c cudaMemPool_t.
+//!
+//! .. warning::
+//!
+//!    ``async_memory_resource`` does not own the pool and it is the responsibility of the user to ensure that the
+//!    lifetime of the pool exceeds the lifetime of the ``async_memory_resource``.
+//!
+//! @endrst
+class async_memory_resource
```

Comment:
I think this name is too generic. This name could indicate just about any asynchronous memory resource, and it's very similar to the concept name `async_resource`.

I think it should at least be `cuda_async_memory_resource`. (Isn't the `cudaMalloc` resource called `cuda_memory_resource`? If the `cudaMalloc` resource were named analogously to this one, it would just be called `memory_resource`!)

Reply:
We actually changed it from `cuda_async_memory_resource` to just `async_memory_resource` because it is in namespace `cuda::mr` and many found that to be too much duplication.

Note that we also renamed the other resource
`cuda::mr::device_memory_resource`
`cuda::mr::managed_memory_resource`
`cuda::mr::pinned_memory_resource`

Reply:
`cuda::mr::device_memory_resource` --> `cudaMalloc`
`cuda::mr::async_memory_resource` --> `cudaMallocAsync`
`cuda::mr::managed_memory_resource` --> `cudaMallocManaged`
`cuda::mr::pinned_memory_resource` --> `cudaHostAlloc`

The first two are inconsistently named.  I also think `device_memory_resource` is confusing because both of the first two above allocate device memory. Arguably the third one does as well.

If we go with `device_memory_resource`, then the cudaMallocAsync one should be `async_device_memory_resource` since they allocate the same type of memory.

Reply:
I dont think that we gain anything by adding more words to the names. The fact that managed or pinned memory is also device accessible does not make it so that memory allocated by cudaMalloc is device only. 

So I believe that `cuda::mr::device_memory_resource` is a fitting name for the simplest way of allocating memory on device.

So far there is no host accessible asynchronous memory and until that is the case I want to keep the default case as simple as possible. We can later add `async_managed_memory_resource` if that ever happens

Reply:
> if that ever happens

It is in the works.

Reply:
> > if that ever happens
> 
> It is in the works.

I know, but my position is still that if we extent functionality we should make that clear and not the default case.

so rather than starting with `cuda::mr::async_device_memory_resource` I would still start with  `cuda::mr::async_memory_resource` and then add `cuda::mr::async_managed_memory_resource` because that is descriptive enough